[ResponseOps][MaintenanceWindow] Public maintenance window API

adcoelho opened the pull request:
Closes #198685

## Summary

This PR creates a public Maintenance Window API.

The work was done on a feature branch over multiple separate PRs. Here we will merge the feature branch into `main`.
- https://github.com/elastic/kibana/pull/209734
- https://github.com/elastic/kibana/pull/213694
- https://github.com/elastic/kibana/pull/214747
- https://github.com/elastic/kibana/pull/213771

The [documentation PR](https://github.com/elastic/kibana/pull/212377) still needs to be merged.

## Release Notes

Publish new public APIs for the Maintenance Window.



<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

adcoelho committed:
[ResponseOps][MW] POST, GET and DELETE maintenance window APIs (#209734)

Closes #208882
Closes #208879
Closes #208880

This PR introduces 3 new public APIs for the maintenance window. GET,
POST, and DELETE maintenance window.

**Please consider the following while reviewing the PR:**
- I wanted the first PR to be just `GET` but that wasn't enough to have
functional tests. This is why I also added POST and DELETE.
- **This branch is being merged into a feature branch**. It will allow
for small changes in the schema until all APIs are done.
- The PR has many changes but most relate to the existing internal API,
which was moved to `/internal`. Those files don't need to be reviewed.
- The `schedule` schema is being developed in a separate PR. Some
changes will come **after** this PR. The places that need to be changed
have been tagged with `// TODO schedule schema`.

- Route locations:
    - `alerting/server/routes/maintenance_window/apis/external/get/`
    - `alerting/server/routes/maintenance_window/apis/external/create/`
    - `alerting/server/routes/maintenance_window/apis/external/delete/`
- Transforms:
-
`alerting/server/routes/maintenance_window/apis/external/common/transforms/`
- there are also route-specific transforms inside
`alerting/server/routes/maintenance_window/apis/external/*`
- Schemas:
- `alerting/common/routes/maintenance_window/external/apis/get/schemas/`
-
`alerting/common/routes/maintenance_window/external/apis/create/schemas/`
-
`alerting/common/routes/maintenance_window/external/apis/delete/schemas/`
- `alerting/common/routes/maintenance_window/external/response/schemas/`
- Types:
- `alerting/common/routes/maintenance_window/external/apis/get/types/`
-
`alerting/common/routes/maintenance_window/external/apis/create/types/`
-
`alerting/common/routes/maintenance_window/external/apis/delete/types/`
- `alerting/common/routes/maintenance_window/external/response/types/`
- Functional tests:
-
`x-pack/test/alerting_api_integration/security_and_spaces/group3/tests/maintenance_window/external/`

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into mw-public-api

adcoelho committed:
[ResponseOps][MW] Public archive maintenance window API (#213771)

Closes #211181

## Summary

This PR introduces a public API for archiving maintenance windows.

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

adcoelho committed:
[ResponseOps][MW] Add custom schedule to GET, POST and DELETE Maintenance Window public APIs (#214747)

Merging into a feature branch.

## Summary

- Add the newly created custom schedule schema to
`createMaintenanceWindowRequestBodySchema` and
`maintenanceWindowResponseSchema`.
- Update `transformInternalMaintenanceWindowToExternal`.
- Update `transformCreateBody`.
- Update tests.

## Testing

The API URL is:
- `https://localhost:5601/api/alerting/maintenance_window`

Here is an example request for creating a maintenance window in Postman:

```
{
    "title": "test-maintenance-window",
    "schedule": {
        "custom": {
            "duration": "1d",
            "start": "1992-01-01T05:00:00.200Z",
            "recurring": {
                "every": "10d",
                "onWeekDay": ["MO", "FR"]
            }
        }
    },
    "scope": {
        "alerting": {
            "query": {
                "kql": "_id: '1234'"
            }
        }
    }
}
```

adcoelho committed:
[ResponseOps][MW] Public PATCH maintenance window (#213694)

Closes #208881

## Merging into a feature branch

## Summary

Public PATCH maintenance window API.

- Partial updates.
- `category_id` is not updated.
- The validation in the maintenance window client that requires scoped
queries to be sent with a `category_id`, has been removed.

## Testing

A Maintenance window has first to be created. It can be done via the UI
or the new public API.

The API URL is:
- `https://localhost:5601/api/alerting/maintenance_window`

Here is an example request for **creating** a maintenance window in
Postman:

```
{
    "title": "test-maintenance-window",
    "schedule": {
        "custom": {
            "duration": "1d",
            "start": "1992-01-01T05:00:00.200Z",
            "recurring": {
                "every": "10d",
                "onWeekDay": ["MO", "FR"]
            }
        }
    },
    "scope": {
        "alerting": {
            "query": {
                "kql": "_id: '1234'"
            }
        }
    }
}
```

The maintenance window can now be updated by sending a patch request to
the same URL with the maintenance window ID appended to it.

- PATCH to
`https://localhost:5601/api/alerting/maintenance_window/<MW_ID>`

The request body is the same as for the POST request. The only
difference is every field is now optional.

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/289347)
* Commit: 50bf568d733b8c70aee04e3d1abbef957f470394

### Failed CI Steps
* [FTR Configs #32](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1305-4c2b-97ae-6929da6543d7)
* [FTR Configs #32](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5c7-83d8-443f-b9ba-b7db5c67e811)
* [FTR Configs #47](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-130f-464b-8801-1d88e7e8468a)
* [FTR Configs #66](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-131c-449d-9f0b-ea8e6b7d4f7e)
* [FTR Configs #96](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1330-4a9a-9f4e-2102c98d5207)
* [FTR Configs #120](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1340-4261-9729-e31a75748e62)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-130f-464b-8801-1d88e7e8468a) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5b0-130f-464b-8801-1d88e7e8468a/artifacts/0195f5d2-ab83-4b72-8a46-13cef56e28fc) FTR Configs #<span></span>47 / aiops change point detection UI detects a change point when no split field is selected
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-131c-449d-9f0b-ea8e6b7d4f7e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5b0-131c-449d-9f0b-ea8e6b7d4f7e/artifacts/0195f5ce-4fd9-4d8a-96c5-ecbbe97bd77d) FTR Configs #<span></span>66 / Cases - group 1 View case Lens visualization adds lens visualization in description
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5c7-83d8-443f-b9ba-b7db5c67e811) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5c7-83d8-443f-b9ba-b7db5c67e811/artifacts/0195f5d3-fb0c-4fcb-92fc-0c84991c1e71) FTR Configs #<span></span>32 / security app Management "after all" hook in "Management"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5c7-83d8-443f-b9ba-b7db5c67e811) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5c7-83d8-443f-b9ba-b7db5c67e811/artifacts/0195f5d3-fb0c-4055-8e4c-f73de1517009) FTR Configs #<span></span>32 / security app Management Security navigation Can navigate to edit user section
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1305-4c2b-97ae-6929da6543d7) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5b0-1305-4c2b-97ae-6929da6543d7/artifacts/0195f5c7-6874-4171-b056-3189c2456732) FTR Configs #<span></span>32 / security app Security Login Page can login
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1330-4a9a-9f4e-2102c98d5207) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5b0-1330-4a9a-9f4e-2102c98d5207/artifacts/0195f5d0-a357-4ca9-b052-2ad6b3554a9d) FTR Configs #<span></span>96 / serverless observability UI Serverless Observability Cases Case View Lens visualization adds lens visualization in description
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1340-4261-9729-e31a75748e62) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5b0-1340-4261-9729-e31a75748e62/artifacts/0195f5d1-5e38-446b-b82d-2b9e3fddc873) FTR Configs #<span></span>120 / Spaces app Copy Saved Objects to Space "before all" hook for "allows a dashboard to be copied to the marketing space, with all references"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5b0-1305-4c2b-97ae-6929da6543d7) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5b0-1305-4c2b-97ae-6929da6543d7/artifacts/0195f5c7-6874-4c9a-b0ed-ca65f3f767fb) FTR Configs #<span></span>32 / X-Pack Accessibility Tests - Group 1 Security Accessibility Login Page User can login with a11y requirements
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/289347#0195f5c7-83d8-443f-b9ba-b7db5c67e811) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/289347/jobs/0195f5c7-83d8-443f-b9ba-b7db5c67e811/artifacts/0195f5d3-fb0c-4e0c-bea3-e008bb9c6e71) FTR Configs #<span></span>32 / X-Pack Accessibility Tests - Group 1 Security Accessibility Login Page User can login with a11y requirements

cc @adcoelho

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289347","number":289347,"commit":"50bf568d733b8c70aee04e3d1abbef957f470394"}],"number":289347}
buildkite-pr-comment-->

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into mw-public-api

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

adcoelho committed:
change maintenance window api url

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/292005)
* Commit: 82f223ad8d95fe1279896cce129f6fc967034009

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/827199b8a1900b7e91cd7c85c04e09abe7003d22) | [after](https://github.com/elastic/kibana/commit/82f223ad8d95fe1279896cce129f6fc967034009) | diff |
| --- | --- | --- | --- |
| `alerting` | 845 | 848 | +3 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/827199b8a1900b7e91cd7c85c04e09abe7003d22) | [after](https://github.com/elastic/kibana/commit/82f223ad8d95fe1279896cce129f6fc967034009) | diff |
| --- | --- | --- | --- |
| `alerting` | 24.2KB | 24.4KB | +209.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/827199b8a1900b7e91cd7c85c04e09abe7003d22) | [after](https://github.com/elastic/kibana/commit/82f223ad8d95fe1279896cce129f6fc967034009) | diff |
| --- | --- | --- | --- |
| `alerting` | 879 | 882 | +3 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/827199b8a1900b7e91cd7c85c04e09abe7003d22) | [after](https://github.com/elastic/kibana/commit/82f223ad8d95fe1279896cce129f6fc967034009) | diff |
| --- | --- | --- | --- |
| `@kbn/test-suites-xpack` | 728 | 734 | +6 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/827199b8a1900b7e91cd7c85c04e09abe7003d22) | [after](https://github.com/elastic/kibana/commit/82f223ad8d95fe1279896cce129f6fc967034009) | diff |
| --- | --- | --- | --- |
| `@kbn/test-suites-xpack` | 754 | 760 | +6 |


</details>


### History
* :broken_heart: [Build #289347](https://buildkite.com/elastic/kibana-pull-request/builds/289347) failed 50bf568d733b8c70aee04e3d1abbef957f470394

cc @adcoelho

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/292005","number":292005,"commit":"82f223ad8d95fe1279896cce129f6fc967034009"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289347","number":289347,"commit":"50bf568d733b8c70aee04e3d1abbef957f470394"}],"number":292005}
buildkite-pr-comment-->

cnasikas reviewed (approved):
Great work 🚀 ! LGTM! I left some minor comments. Also: 

- The public routes do not have any tag like we have for alerting. Should we add `tags: ['oas-tag:maintenance_window']`? @lcawl Wdyt? We want a dedicated category, outside of alerting, for the new public Maintenance Window APIs/

- Some routes are missing the `400` in the response. Should we add it?

- A disabled MW is shown as running in the UI.  Unrelated to the PR. Could you please create an issue and link it to the MW GA meta issue?

- If I set `every` more that `1d` the `onWeekDay` (day selection) is not shown in the UI. Unrelated to the PR. Could you please create an issue and link it to the MW GA meta issue?

- If I create a MW from the UI I can get: "duration": "2.0208333333333335d". Unrelated to the PR. Could you please create an issue to always return an integer (to the first unit that you can do it) and link it to the MW GA meta issue?

- From the UI I can create a MW with `duration` > `every` but from the APIs I cannot. I remember deciding that but could you please remind me why?

adcoelho committed:
Addresses PR comments.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into mw-public-api

adcoelho commented:
Thanks for the review @cnasikas 👍 

> The public routes do not have any tag like we have for alerting. Should we add tags: ['oas-tag:maintenance_window']? @lcawl Wdyt? We want a dedicated category, outside of alerting, for the new public Maintenance Window APIs/

We can wait for lisa's comment but I added the tags anyway for now. It seems like a reasonable suggestion.

> Some routes are missing the 400 in the response. Should we add it?

Added.

> A disabled MW is shown as running in the UI. Unrelated to the PR. Could you please create an issue and link it to the MW GA meta issue?
> If I set every more that 1d the onWeekDay (day selection) is not shown in the UI. Unrelated to the PR. Could you please create an issue and link it to the MW GA meta issue?
> If I create a MW from the UI I can get: "duration": "2.0208333333333335d". Unrelated to the PR. Could you please create an issue to always return an integer (to the first unit that you can do it) and link it to the MW GA meta issue?

#218109
#218111
#218113

> From the UI I can create a MW with duration > every but from the APIs I cannot. I remember deciding that but could you please remind me why?

Maybe @js-jankisalvi can chip in here? The API validation comes from the custom schedule.

js-jankisalvi commented:
> > From the UI I can create a MW with duration > every but from the APIs I cannot. I remember deciding that but could you please remind me why?
> 
> Maybe @js-jankisalvi can chip in here? The API validation comes from the custom schedule.

This is because creating MW from UI still uses internal api `x-pack/platform/plugins/shared/alerting/public/services/maintenance_windows_api/create.ts` which uses `rRuleRequestSchemaV1` instead of `scheduleRequestSchemaV1`.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/292742)
* Commit: a9b8f98be471800de3c631bfc25800147f8ccfe9

### Failed CI Steps
* [Jest Tests #6](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452)
* [Jest Tests #6](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/01963486-f3b2-42b8-930e-95f61d133452/artifacts/019634af-030b-4e09-be4c-bd3492bccf58) Jest Tests #<span></span>6 / archiveMaintenanceWindowRoute should archive the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/019634af-4196-4282-94e7-cf9489f20205/artifacts/019634bf-d241-4cb9-a631-a596fea63518) Jest Tests #<span></span>6 / archiveMaintenanceWindowRoute should archive the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/01963486-f3b2-42b8-930e-95f61d133452/artifacts/019634af-030a-411f-be85-2bed3306166a) Jest Tests #<span></span>6 / createMaintenanceWindowRoute should create the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/019634af-4196-4282-94e7-cf9489f20205/artifacts/019634bf-d243-4063-a17b-e6f754c9070a) Jest Tests #<span></span>6 / createMaintenanceWindowRoute should create the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/01963486-f3b2-42b8-930e-95f61d133452/artifacts/019634af-030d-48f0-9835-f8d5f83a73e8) Jest Tests #<span></span>6 / deleteMaintenanceWindowRoute should delete the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/019634af-4196-4282-94e7-cf9489f20205/artifacts/019634bf-d240-452c-95ae-4c1d00ac17e3) Jest Tests #<span></span>6 / deleteMaintenanceWindowRoute should delete the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/01963486-f3b2-42b8-930e-95f61d133452/artifacts/019634af-030c-4688-b039-bf39a777bf93) Jest Tests #<span></span>6 / getMaintenanceWindowRoute should get the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/019634af-4196-4282-94e7-cf9489f20205/artifacts/019634bf-d244-4b61-8e43-82a294925509) Jest Tests #<span></span>6 / getMaintenanceWindowRoute should get the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/01963486-f3b2-42b8-930e-95f61d133452/artifacts/019634af-030c-4960-8fc0-b400ad6a169c) Jest Tests #<span></span>6 / unarchiveMaintenanceWindowRoute should unarchive the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/019634af-4196-4282-94e7-cf9489f20205/artifacts/019634bf-d243-4fe4-a8e9-fd4fd7a0fe2a) Jest Tests #<span></span>6 / unarchiveMaintenanceWindowRoute should unarchive the maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#01963486-f3b2-42b8-930e-95f61d133452) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/01963486-f3b2-42b8-930e-95f61d133452/artifacts/019634af-030b-4d99-a48b-4c09f2a0ded5) Jest Tests #<span></span>6 / updateMaintenanceWindowRoute should update a maintenance window
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/292742#019634af-4196-4282-94e7-cf9489f20205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/292742/jobs/019634af-4196-4282-94e7-cf9489f20205/artifacts/019634bf-d243-48c9-bb01-1093dec197bf) Jest Tests #<span></span>6 / updateMaintenanceWindowRoute should update a maintenance window

### History
* :green_heart: [Build #292005](https://buildkite.com/elastic/kibana-pull-request/builds/292005) succeeded 82f223ad8d95fe1279896cce129f6fc967034009
* :broken_heart: [Build #289347](https://buildkite.com/elastic/kibana-pull-request/builds/289347) failed 50bf568d733b8c70aee04e3d1abbef957f470394

cc @adcoelho

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/292742","number":292742,"commit":"a9b8f98be471800de3c631bfc25800147f8ccfe9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/292005","number":292005,"commit":"82f223ad8d95fe1279896cce129f6fc967034009"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289347","number":289347,"commit":"50bf568d733b8c70aee04e3d1abbef957f470394"}],"number":292742}
buildkite-pr-comment-->

lcawl commented on x-pack/platform/plugins/shared/alerting/server/routes/maintenance_window/apis/external/delete/delete_maintenance_window_route.test.ts:
```diff
@@ -0,0 +1,120 @@
+    expect(config.options).toMatchInlineSnapshot(`
+      Object {
+        "access": "public",
+        "summary": "Deletes a maintenance window by ID.",
```
Unless there's another endpoint that deletes a single maintenance window by another type of identifier, this can be shortened:
```suggestion
        "summary": "Delete a maintenance window.",
```

lcawl commented on x-pack/platform/plugins/shared/alerting/server/routes/maintenance_window/apis/external/get/get_maintenance_window_route.test.ts:
```diff
@@ -0,0 +1,140 @@
+    expect(config.options).toMatchInlineSnapshot(`
+      Object {
+        "access": "public",
+        "summary": "Gets a maintenance window by ID.",
```
```suggestion
        "summary": "Get maintenance window details.",
```

lcawl commented on x-pack/platform/plugins/shared/alerting/server/routes/maintenance_window/apis/external/get/get_maintenance_window_route.ts:
```diff
@@ -0,0 +1,82 @@
+      },
+      options: {
+        access: 'public',
+        summary: 'Gets a maintenance window by ID.',
```
```suggestion
        summary: 'Get maintenance window details.',
```

lcawl commented on x-pack/platform/plugins/shared/alerting/server/routes/maintenance_window/apis/external/unarchive/unarchive_maintenance_window_route.test.ts:
```diff
@@ -0,0 +1,156 @@
+    expect(config.options).toMatchInlineSnapshot(`
+      Object {
+        "access": "public",
+        "summary": "Unarchives a maintenance window by ID.",
```
```suggestion
        "summary": "Unarchive a maintenance window.",
```

lcawl commented on x-pack/platform/plugins/shared/alerting/server/routes/maintenance_window/apis/external/unarchive/unarchive_maintenance_window_route.ts:
```diff
@@ -0,0 +1,82 @@
+      },
+      options: {
+        access: 'public',
+        summary: 'Unarchives a maintenance window by ID.',
```
```suggestion
        summary: 'Unarchive a maintenance window.',
```

lcawl reviewed (commented):
Thanks for adding the `oas-tag`, it is indeed necessary for that info to show up properly in the docs.

IMO in addition to the operation summary, an operation description will ultimately be necsesary for the "archive" and "unarchive" endpoints, since it's not obvious what that accomplishes.

adcoelho committed:
Fixes tests

adcoelho commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/293118)
* Commit: ac14e76cc0b965c35163d6cd288fffeb5dcf2b22

### Failed CI Steps
* [FTR Configs #49](https://buildkite.com/elastic/kibana-pull-request/builds/293118#0196390e-e50c-4e07-bd8a-db19d399414f)
* [FTR Configs #49](https://buildkite.com/elastic/kibana-pull-request/builds/293118#0196393b-4a81-4e40-a282-b9970b8be53b)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293118#0196390e-e50c-4e07-bd8a-db19d399414f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293118/jobs/0196390e-e50c-4e07-bd8a-db19d399414f/artifacts/0196393a-e053-47b6-8d89-ea049de57af3) FTR Configs #<span></span>49 / console app misc console behavior keyboard shortcuts open documentation should open documentation when Ctrl+/ is pressed
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293118#0196393b-4a81-4e40-a282-b9970b8be53b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293118/jobs/0196393b-4a81-4e40-a282-b9970b8be53b/artifacts/01963946-9fe5-4613-b2db-20017fe9c2df) FTR Configs #<span></span>49 / console app misc console behavior keyboard shortcuts open documentation should open documentation when Ctrl+/ is pressed

### History
* :broken_heart: [Build #292742](https://buildkite.com/elastic/kibana-pull-request/builds/292742) failed a9b8f98be471800de3c631bfc25800147f8ccfe9
* :green_heart: [Build #292005](https://buildkite.com/elastic/kibana-pull-request/builds/292005) succeeded 82f223ad8d95fe1279896cce129f6fc967034009
* :broken_heart: [Build #289347](https://buildkite.com/elastic/kibana-pull-request/builds/289347) failed 50bf568d733b8c70aee04e3d1abbef957f470394

cc @adcoelho

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293118","number":293118,"commit":"ac14e76cc0b965c35163d6cd288fffeb5dcf2b22"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/292742","number":292742,"commit":"a9b8f98be471800de3c631bfc25800147f8ccfe9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/292005","number":292005,"commit":"82f223ad8d95fe1279896cce129f6fc967034009"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289347","number":289347,"commit":"50bf568d733b8c70aee04e3d1abbef957f470394"}],"number":293118}
buildkite-pr-comment-->

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into mw-public-api

adcoelho committed:
[ResponseOps][MW] Maintenance Window oas documentation  (#212377)

### Merging into a feature branch

## Summary

I recently added a few public routes for managing maintenance windows. I
will need to generate documentation for these routes for my' on week'
task.

- Updated `capture_oas_snapshot` to include
`/api/alerting/maintenance_window`
- Added `tags` to the maintenance window routes

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: lcawl <lcawley@elastic.co>

adcoelho committed:
Merge branch 'main' into mw-public-api

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/14489613602

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.x|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 216756
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
